### PR TITLE
[Feature] Add AutoSuggestion option for TextBox

### PR DESCRIPTION
**Pull request by AbderrahmaneAhmam**
### Description:
Hello everyone,

This pull request introduces a new feature to the MaterialDesignInXamlToolkit project, enabling the AutoSuggestion option for the TextBox control. This feature enhances user experience by providing automatic suggestions as the user types. The search algorithm is controlled by the developer in the ViewModel part, making it customizable and flexible.

### Key Changes:

- Added AutoSuggestionEnabled property to the TextFieldAssist.
- Added AutoSuggestionItemsSource property to the TextFieldAssist.
- Added AutoSuggestionItemTemplate property to the TextFieldAssist.
- Added AutoSuggestionItemTemplateSelector property to the TextFieldAssist.
- Added AutoSuggestionValueMember property to the TextFieldAssist.
- Added AutoSuggestionDisplayMember property to the TextFieldAssist.

This contribution aligns with the MaterialDesignInXamlToolkit project's goals of providing modern and user-friendly UI components. As a first-time contributor to an open source project, I am thrilled to be a part of this community and contribute to its growth. I would greatly appreciate any feedback and guidance to improve the feature and align it with the project's standards.

I have thoroughly tested the feature and ensured it seamlessly integrates with the existing codebase. I'm excited about this contribution and believe it will be valuable for the project's users.

I'm looking forward to your feedback and the opportunity to collaborate on further enhancing this feature.

Please let me know if there are any remarks or changes you would like me to make.

### Screenshots & Demo
Demo_AutoSuggestion_AdobeExpress

### CODE REVIEW

**Review by Keboo — CHANGES_REQUESTED** [review]
Hi @AbderrahmaneAhmam

Thank you for this. I think this would be a nice addition as there have been requests for something similar in the past. 

A few suggestions before we merge it. 
1) I think this would be better as a new control that derives from TextBox, rather than adding a bunch of attached properties. This would move all of the attached properties into dependency properties on the new control. 
2) This will likely mean duplicating the existing TextBox Styles/Templates to target the new control type.
3) I think doing this would allow for a cleaner setup. The needed setup right now needing the register for the TextChanged to show the drop down is a bit awkward to use. With a custom control, it would be able to register its own TextChanged event handler to manage showing the list.

Again, good work so far.

File: MainDemo.Wpf/Domain/FieldsViewModel.cs [hunk 1]
```diff
@@ -69,15 +75,81 @@ public class FieldsViewModel : ViewModelBase
...
+            set { SetProperty(ref baseListSuggestion, value); }
+        }
+
+        private ObservableCollection<KeyValuePair<string, Brush>> listColors;
+
+        public ObservableCollection<KeyValuePair<string, Brush>> ListColors
+        {
+            get { return listColors; }
+            set { SetProperty(ref listColors, value); }
+        }
```
[the remark this answers is not in the document]
Comment (AbderrahmaneAhmam) [reply MainDemo.Wpf/Domain/FieldsViewModel.cs]: Thanks for your reply.
Is it acceptable to duplicate all the styles of the TextBox to be reused in the custom control? If so, please provide a suggestion for the name of the custom control. Additionally, for the TextChanged event, I propose extracting it as a command. This approach allows for better control of the search algorithm through the ViewModel.

[the remark this answers is not in the document]
Comment (AbderrahmaneAhmam) [reply MainDemo.Wpf/Domain/FieldsViewModel.cs]: Thanks @Keboo. That's clear. I will start working on this task and try to finish it as soon as possible.

[the remark this answers is not in the document]
Comment (Keboo) [reply MainDemo.Wpf/Domain/FieldsViewModel.cs]: Feel free to reach out if you have questions. I really appreciate the help and look forward to it.

[the remark this answers is not in the document]
Comment (AbderrahmaneAhmam) [reply MainDemo.Wpf/Domain/FieldsViewModel.cs]: Hello @Keboo ,

I have implemented this approach. Could you please verify if any adjustments or modifications are needed? I have not yet completed the testing phase, and for testing purposes, I have used the same file "Fields.Xaml". I am unsure if we need to create a separate page for this new control.

Thank you for your assistance.

**Comment by AbderrahmaneAhmam** [comment]
Hello @Keboo,

I am currently working on the TODOs, but I am having difficulty understanding this specific part:
"TODO: Test case with ICollectionSource." Could you please provide some clarification?

**Comment by Keboo** [comment]
> Hello @Keboo,
> 
> I am currently working on the TODOs, but I am having difficulty understanding this specific part: "TODO: Test case with ICollectionSource." Could you please provide some clarification?

Great. Sorry for just jumping in and doing some work on it, kinda got a little over excited and I knew my recent file-scope namespace all the things was going to cause conflicts. 


For the `ICollectionView` I have pushed another commit that shows what I was thinking (the test is currently failing).
I suspect that to get the selection working, we may end up needing to make the base class `Selector` and then add additional properties to support the normal TextBox behavior (In this case I think the behavior of this control may be closer to a `ComboBox` than a `TextBox`).

**Comment by AbderrahmaneAhmam** [comment]
> > Hello @Keboo,
> > I am currently working on the TODOs, but I am having difficulty understanding this specific part: "TODO: Test case with ICollectionSource." Could you please provide some clarification?
> 
> Great. Sorry for just jumping in and doing some work on it, kinda got a little over excited and I knew my recent file-scope namespace all the things was going to cause conflicts.
> 
> For the `ICollectionView` I have pushed another commit that shows what I was thinking (the test is currently failing). I suspect that to get the selection working, we may end up needing to make the base class `Selector` and then add additional properties to support the normal TextBox behavior (In this case I think the behavior of this control may be closer to a `ComboBox` than a `TextBox`).

Hi @Keboo 

Currently, we have the capability to extract these properties from the ListBox utilized within this control. The principle behind this is to construct a control similar to a TextBox, enabling its use alongside other TextBox features. The sole distinguishing feature is the auto-suggestion functionality.

I've conducted research on similar controls, and they function using a similar approach. The base class is a TextBox. For instance, let's consider Syncfusion's [Autocomplete](https://help.syncfusion.com/wpf/autocomplete/getting-started): the base class of SfTextBoxExt is a TextBox. However, it offers additional features like SelectedItem and other properties from the Selector class.


```csharp
public class SfTextBoxExt : TextBox, IDisposable, ITextInputLayoutSelector
```

```csharp
public object SelectedItem { get; set; }
public IEnumerable SelectedItems { get; set; }
public object SelectedValue { get; set; }
```

**Comment by AbderrahmaneAhmam** [comment]
Hi @Keboo,

sorry, I understand now what you are referring to with "IsSynchronizedWithCurrentItem". To be honest, I haven't used it before, but after doing some research, I've gained a better understanding. I have a suggestion: how about enabling it by default in the ListBox of AutoSuggestBox? This would allow us to use it for keyboard navigation through ICollectionView. Additionally, this option would be available for consumers as well.

**Comment by Keboo** [comment]
> Hi @Keboo,
> 
> sorry, I understand now what you are referring to with "IsSynchronizedWithCurrentItem". To be honest, I haven't used it before, but after doing some research, I've gained a better understanding. I have a suggestion: how about enabling it by default in the ListBox of AutoSuggestBox? This would allow us to use it for keyboard navigation through ICollectionView. Additionally, this option would be available for consumers as well.

I think that would be a good idea.

**Comment by anno0** [comment]
Will it be possible to use it in a RichTextBox?

**Comment by AbderrahmaneAhmam** [comment]
> Will it be possible to use it in a RichTextBox?

Actually, this pull request is not related to the RichTextBox; it's about creating a new control, the AutoSuggestBox.

**Review by Keboo — APPROVED** [review]

**Pull request merged**